### PR TITLE
[#643] Tests for AMQP Adapter

### PR DESCRIPTION
**Pull request by Alfusainey**
@sophokles73: This PR test the functionality of the AMQP adapter (unauthenticated devices)

### CODE REVIEW

File: adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterWithoutClientTest.java [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+import io.vertx.proton.ProtonReceiver;
+import io.vertx.proton.ProtonServer;
+
+/**
+ * Verifies the behaviour of {@link VertxBasedAmqpProtocolAdapter}. In this test setup, the AMQP adapter is not deployed
+ * as a running Verticle and does not use an AMQP client to simulate a client device.
+ * Rather, the setup manually instantiates the adapter and gives it a user defined ProtonServer.
+ */
+@RunWith(VertxUnitRunner.class)
+public class VertxBasedAmqpProtocolAdapterWithoutClientTest {
```
Comment (sophokles73) [inline adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterWithoutClientTest.java]: This seems to be the unit tests, right? If so, can we please simply name this class `VertxBasedAmqpProtocolAdapterTest`?

File: adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterWithClientTest.java [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+import io.vertx.proton.ProtonHelper;
+import io.vertx.proton.ProtonQoS;
+import io.vertx.proton.ProtonSender;
+
+/**
+ * Verifies the behaviour of {@link VertxBasedAmqpProtocolAdapter}. In this setup, the AMQP adapter is deployed and run
+ * as a Verticle. A ProtonClient, which simulates a device, is used to communicate with the adapter.
+ */
+@RunWith(VertxUnitRunner.class)
+public class VertxBasedAmqpProtocolAdapterWithClientTest {
```
Comment (sophokles73) [inline adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterWithClientTest.java]: These seem to be functional tests running against a deployed version of the adapter. I would rather move these tests to the `tests` module and run the tests against the Docker containers we start there. It would probably also make more sense to put it into a separate PR then...

File: adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterWithoutClientTest.java (hunk 1 above)
Comment (Alfusainey) [reply adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterWithoutClientTest.java]: yes, it is. i rename it to distinguish it from the other test. will change it now

File: adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterWithClientTest.java (hunk 2 above)
Comment (Alfusainey) [reply adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterWithClientTest.java]: ok, i see. thanks!

**Comment by Alfusainey** [comment]
@sophokles73: PR updated!

File: adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterTest.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    private Message getFakeMessage() {
+        final Message message = mock(Message.class);
+        when(message.getContentType()).thenReturn("application/text");
+        when(message.getBody()).thenReturn(new AmqpValue("some payload"));
+        return message;
+    }
+
+    private MessageSender givenAmessageSenderForAnyTenant() {
```
Comment (sophokles73) [inline adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterTest.java]: this should better be called *givenATelemetrySenderForAnyTenant* in order to point out that the sender indeed is a sender for telemetry messages.

File: adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterTest.java [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+     * Verifies that a request to upload telemetry message (with settled=true) results in the sender sending the message
+     * without waiting for a response from the downstream peer.
+     * 
+     * AT_MOST_ONCE delivery semantics
+     */
+    @Test
+    public void uploadTelemetryMessageWithSettledDeliverySemantics() {
+        // GIVEN an AMQP adapter with a configured server
+        final VertxBasedAmqpProtocolAdapter adapter = givenAnAmqpAdapter();
+        final MessageSender sender = givenAmessageSenderForAnyTenant();
```
Comment (sophokles73) [inline adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterTest.java]: better call this variable *telemetrySender*

File: adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterTest.java [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    /**
+     * Verifies that a request to upload telemetry message (with settled=false) results in the sender sending the
+     * message and waits for a response from the downstream peer.
+     * 
+     * AT_LEAST_ONCE delivery semantics.
+     */
+    @Test
+    public void uploadTelemetryMessageWithNotSettledDeliverySemantics() {
```
Comment (sophokles73) [inline adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterTest.java]: UnsettledDeliverySemantics

File: adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterTest.java [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        // GIVEN an adapter configured to use a user-define server.
+        final VertxBasedAmqpProtocolAdapter adapter = givenAnAmqpAdapter();
+        final MessageSender sender = givenAmessageSenderForAnyTenant();
```
Comment (sophokles73) [inline adapters/amqp-vertx/src/test/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapterTest.java]: telemetrySender

**Comment by Alfusainey** [comment]
@sophokles73: thanks, just updated the PR!

**Comment by Alfusainey** [comment]
thanks. i've squashed the commits into one
